qualtrics_link/management/commands/qualtrics.py

```diff
@@ -334,93 +306,121 @@ def filter_users(slice_amount):
-                        'role': update_dict.get('role', p_role),
+                matching_person = None
+
+                # If we have a QualtricsUser record, then we can use the univ_id from that to get a Person record rather
+                # than having to do other comparisons to find a matching person.
+                if qualtrics_user:
+                    person_queryset = Person.objects.filter(univ_id=qualtrics_user.univ_id)
+                    if len(person_queryset) > 0:
+                        matching_person = person_queryset[0]
+                else:
+                    # Filter Person records based off of their email if present
+                    if q_email:
+                        person_queryset = Person.objects.filter(email_address=q_email)
+                        # A user may have had their Qualtrics account created with a different email address than what
+                        # is currently stored in our DB. If so, we need to look them up based on first and last name.
+                        if len(person_queryset) == 0:
+                            person_queryset = Person.objects.filter(name_first=q_first_name, name_last=q_last_name)
+                    else:
+                        person_queryset = Person.objects.filter(name_first=q_first_name, name_last=q_last_name)
+
+                    # Iterate through the queryset, encrypting the current Persons HUID to find a match with the q_id
+                    for person in person_queryset:
+                        enc_id = util.get_encrypted_huid(person.univ_id) + '#harvard'
+                        if enc_id == q_username:
+                            matching_person = person
+                            break
+
+                # Compare the Qualtrics data against the LDAP People data
+                if matching_person is not None:
+                    person_details = util.get_person_details(matching_person.univ_id, person_queryset)
+                    p_role = user_type_mapping[person_details.role]
+                    p_division = division_mapping[person_details.division]
+
+                    update_person = False
+                    update_dict = {
+                        'user_id': q_id
                     }
+                    # Do the comparison of what is currently in Qualtrics and what the matching person has
+                    # If they differ, then add to the filtered list the current persons q_id, role and division
+                    if p_division != q_division:
+                        update_dict['division'] = p_division
+                        update_person = True
+
+                    # If the Person is not a Student or Employee, or Brand Admin then they need to be updated
+                    # Or if their roles do not match
+                    if q_role not in user_type_list or p_role != q_role:
+                        update_dict['user_type'] = p_role
+                        update_person = True
```

Comment:
It  it worth considering the addition  of expiration date /role_end_date check with the current discussion of role end date?  
@cmurtaugh : thoughts?

Reply:
Jill is creating a separate story to address this issue